linkerd/meshtls/boring/src/server.rs

```diff
@@ -0,0 +1,177 @@
+use crate::creds::CredsRx;
+use linkerd_identity::Name;
+use linkerd_io as io;
+use linkerd_stack::{Param, Service};
+use linkerd_tls::{ClientId, LocalId, NegotiatedProtocol, ServerTls};
+use std::{future::Future, pin::Pin, sync::Arc, task::Context};
+use tracing::debug;
+
+#[derive(Clone)]
+pub struct Server {
+    name: Name,
+    rx: CredsRx,
+    alpn: Option<Arc<[Vec<u8>]>>,
+}
+
+pub type TerminateFuture<I> =
+    Pin<Box<dyn Future<Output = io::Result<(ServerTls, ServerIo<I>)>> + Send>>;
+
+#[derive(Debug)]
+pub struct ServerIo<I>(tokio_boring::SslStream<I>);
+
+// === impl Server ===
+
+impl Server {
+    pub(crate) fn new(name: Name, rx: CredsRx) -> Self {
+        Self {
+            name,
+            rx,
+            alpn: None,
+        }
+    }
+
+    pub fn with_alpn(mut self, alpn_protocols: Vec<Vec<u8>>) -> Self {
+        if !alpn_protocols.is_empty() {
+            self.alpn = Some(alpn_protocols.into());
+        }
+
+        self
+    }
+}
+
+impl Param<LocalId> for Server {
+    fn param(&self) -> LocalId {
+        LocalId(self.name.clone())
+    }
+}
+
+impl<I> Service<I> for Server
+where
+    I: io::AsyncRead + io::AsyncWrite + Send + Unpin + 'static,
+{
+    type Response = (ServerTls, ServerIo<I>);
+    type Error = std::io::Error;
+    type Future = TerminateFuture<I>;
+
+    #[inline]
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> io::Poll<()> {
+        io::Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, io: I) -> Self::Future {
+        // TODO(ver) we should avoid creating a new context for each connection.
+        let acceptor = match &self.alpn {
+            Some(alpn) => self.rx.borrow().acceptor(alpn),
+            None => self.rx.borrow().acceptor(&[]),
+        };
```

Comment:
nit/TIOLI: i think this could be
```suggestion
        let acceptor = self.rx.borrow().acceptor(self.alpn.as_deref().unwrap_or(&[]);
```